lib/preface_make/apply.ml

```diff
@@ -0,0 +1,185 @@
+open Preface_core.Fun
+
+module Core_via_map_and_product (Req : Preface_specs.Apply.WITH_MAP_AND_PRODUCT) =
+struct
+  include Req
+
+  let apply f a = map (fun (f, a) -> f a) @@ product f a
+  let lift2 f x y = apply (map f x) y
+end
+
+module Core_over_functor_via_apply
+    (Req : Preface_specs.Apply.WITH_APPLY)
+    (Functor : Preface_specs.Functor.WITH_MAP with type 'a t = 'a Req.t) =
+struct
+  include Req
+  include Functor
+
+  let product a b = apply (map (fun a b -> (a, b)) a) b
+  let lift2 f x y = apply (map f x) y
+end
+
+module Core_over_functor_via_lift2
+    (Req : Preface_specs.Apply.WITH_LIFT2)
+    (Functor : Preface_specs.Functor.WITH_MAP with type 'a t = 'a Req.t) =
+struct
+  include Req
+  include Functor
+
+  let apply f a = lift2 (fun x -> x) f a
+  let product a b = apply (map (fun a b -> (a, b)) a) b
+end
```

Comment:
The API is not consistent, CF my comment below.

Reply:
Done